frontend/src/app/main/data/workspace/libraries.cljs

```diff
@@ -716,6 +756,48 @@
                   :callback do-dismiss}]
                 :sync-dialog))))))
 
+(defn watch-component-changes
+  "Watch the state for changes that affect to any main instance. If a change is detected will throw
+  an update-component-sync, so changes are immediately propagated to the component and copies."
+  []
+  (ptk/reify ::watch-component-changes
+    ptk/WatchEvent
+    (watch [_ state stream]
+      (let [components-v2 (features/active-feature? state :components-v2)
+
+            stopper
+            (->> stream
+                 (rx/filter #(or (= :app.main.data.workspace/finalize-page (ptk/type %))
+                                 (= ::watch-component-changes (ptk/type %)))))
+
+            workspace-data-str
+            (->> (rx/concat
+                  (rx/of nil)
+                  (rx/from-atom refs/workspace-data {:emit-current-value? true})))
+
+            change-str
+            (->> stream
+                 (rx/filter #(or (dch/commit-changes? %)
+                                 (= (ptk/type %) :app.main.data.workspace.notifications/handle-file-change)))
+                 (rx/observe-on :async))
+
+            check-changes
+            (fn [[event data]]
+              (let [changes (-> event deref :changes)
+                    components-changed (reduce #(into %1 (ch/components-changed data %2))
+                                               #{}
+                                               changes)]
+                (when (d/not-empty? components-changed)
+                  (apply st/emit!
```

Comment:
I think `run!` is better here than `apply`. I mean

```clojure
(->> components-changed
     (map #(update-component-sync % (:id data)))
     (run! st/emit!))
```

Reply:
Perhaps is clearer and more appropriate, yes.

Althoug the pattern `(apply st/emit! (map xxxx))` is used in many other places.

Reply:
Is used in many sites yes, but this is not the optimal one.